Source/WebCore/page/ContextMenuController.cpp

```diff
@@ -998,29 +1006,38 @@ void ContextMenuController::populate()
                 appendItem(StopItem, m_contextMenu.get());
                 appendItem(ReloadItem, m_contextMenu.get());
 #else
-                if (frame->page() && frame->page()->backForward().canGoBackOrForward(-1))
-                    appendItem(BackItem, m_contextMenu.get());
-
-                if (frame->page() && frame->page()->backForward().canGoBackOrForward(1))
-                    appendItem(ForwardItem, m_contextMenu.get());
-
-                // use isLoadingInAPISense rather than isLoading because Stop/Reload are
-                // intended to match WebKit's API, not WebCore's internal notion of loading status
-                if (loader.documentLoader()->isLoadingInAPISense())
-                    appendItem(StopItem, m_contextMenu.get());
-                else
-                    appendItem(ReloadItem, m_contextMenu.get());
+                if (frame->isMainFrame() || (frame->page() && !frame->ownerElement()->document().isPDFDocument())) {
+                    if (frame->page() && frame->page()->backForward().canGoBackOrForward(-1))
+                        appendItem(BackItem, m_contextMenu.get());
+
+                    if (frame->page() && frame->page()->backForward().canGoBackOrForward(1))
+                        appendItem(ForwardItem, m_contextMenu.get());
+                    
+                    // use isLoadingInAPISense rather than isLoading because Stop/Reload are
+                    // intended to match WebKit's API, not WebCore's internal notion of loading status
+                    if (loader.documentLoader()->isLoadingInAPISense())
+                        appendItem(StopItem, m_contextMenu.get());
+                    else
+                        appendItem(ReloadItem, m_contextMenu.get());
+                }
 #endif
                 }
 
-                if (frame->page() && !frame->isMainFrame())
+                if (frame->page() && !frame->isMainFrame() && !frame->ownerElement()->document().isPDFDocument()) 
```

Comment:
What guarantees that frame->ownerElement() is non null?

Reply:
I think I had some misunderstandings between frames and pages. I will definitely need to add a null check for that. If we got rid of the frame->page() check here would that be ok? Since a frame could get detached from a page (although I don't entirely know what that scenario would be), would we want to check that or do you not think it would matter in this case? Also, it looks like I might want to add a null check for the document as well?

Reply:
> might want to add a null check for the document as well?
No, document() is a reference, not a pointer so it cannot be null.

But when you dereference this pointer (frame->ownerElement()) without a null check, this looks dangerous.

Reply:
Added the frame->ownerElement null pointer check here